worker/janitor.go

```diff
@@ -0,0 +1,181 @@
+package worker
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/WatchBeam/redutil/heartbeat"
+	"github.com/WatchBeam/redutil/queue"
+	"github.com/benbjohnson/clock"
+	"github.com/garyburd/redigo/redis"
+	"github.com/hjr265/redsync.go/redsync"
+)
+
+// The Janitor is used to assist in the tear down of dead workers. It can
+// be provided to the worker to hook additional functionality that will
+// occur when the worker dies.
+type Janitor interface {
+	// Called when a worker dies after we have acquired a lock and before
+	// we start moving the worker's queue back to the main processing
+	// queue. Note that if we does before the worker's queue is moved
+	// over, this function *can be called multiple times on the
+	// same worker*
+	//
+	// If an error is returned from the function, the queue concatenation
+	// will be aborted and we'll release the lock.
+	OnPreConcat(cnx redis.Conn, worker string) error
+
+	// Called when a worker dies after we have acquired a lock and finished
+	// moving the worker's queue back to the main processing queue. Note
+	// that, in the result of a panic or power failure, this function
+	// may never be called, and errors resulting from this function
+	// will not roll-back the concatenation.
+	OnPostConcat(cnx redis.Conn, worker string) error
+}
+
+// Base janitor used unless the user provides a replacement.
+type nilJanitor struct{}
+
+var _ Janitor = nilJanitor{}
+
+func (n nilJanitor) OnPreConcat(cnx redis.Conn, worker string) error  { return nil }
+func (n nilJanitor) OnPostConcat(cnx redis.Conn, worker string) error { return nil }
+
+// The janitor is responsible for cleaning up dead workers.
+type janitorRunner struct {
+	// pool is a *redis.Pool used to maintain and use connections into
+	// Redis.
+	pool *redis.Pool
+
+	availableTasks queue.Queue
+
+	// Associated heartbeater detector
+	detector heartbeat.Detector
+	janitor  Janitor
+
+	// Duration between dead checks. The first check will come at a time
+	// between 0 and time.Duration, so that workers started at the same time
+	// don't try to contest the same locks.
+	interval time.Duration
+
+	// interval checker used to prune dead workers
+	clock clock.Clock
```

Comment:
Any reason we're using a clock here? A `time.Ticker` should work fine, unless I'm missing something.